### PR TITLE
I hate colexigraphical order

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #166657
* __->__ #166605
* #166359

# Summary

Okay so whle testing the mask-mod support and block-mask support I added doc_ids for testing. And noticed the numerics were off.

What was happend. By default ops.load uses the linearized form of indexing. And cutedsl tensor indexing also allows for linearized form. HOWEVER, cute is column major and triton / inductor is row-major. Thus we were incorrectly indexing in the wrong order.

So what does this do?

1. it adds some utilities which make the output code a little easier to understand.
2. It goes from linear index back to the individual dim index and then lets cute handle the majorness

### CODE REVIEW

File: test/inductor/test_flex_flash.py [hunk 1]
```diff
@@ -329,6 +329,40 @@ def custom_mask(b, h, q_idx, kv_idx):
...
+        document_ids = []
+        for lengths in lengths_per_batch:
+            assert sum(lengths) == seq_len
+            doc_tokens = []
+            for doc_id, length in enumerate(lengths):
+                doc_tokens.extend([doc_id] * length)
+            document_ids.append(doc_tokens)
+        document_ids = torch.tensor(document_ids, device=device, dtype=torch.long)
+
+        print("shape of document_ids:", document_ids.shape)
```
Comment (drisspg) [inline test/inductor/test_flex_flash.py:353]: remove

File: torch/_inductor/codegen/cutedsl/cutedsl_kernel.py [hunk 2]
```diff
@@ -483,6 +483,73 @@ def load(self, name: str, index: sympy.Expr):
...
+        Workaround for lack of gather support: SSA values cannot be used directly
+        as indices. This generates code to convert SSA → indexable scalar.
+        """
+        result = self.kernel.cse.newvar(dtype=torch_dtype)
+        self.kernel.body.writeline(
+            f"{result} = ssa_to_indexable({expr_str}, {cute_dtype})"
+        )
+        return str(result)
+
+    def _try_get_structured_indices(
```
Comment (drisspg) [inline torch/_inductor/codegen/cutedsl/cutedsl_kernel.py:501]: this was vibecodee, I need to proofread but so far it works

File: torch/_inductor/codegen/cutedsl/cutedsl_kernel.py [hunk 3]
```diff
@@ -483,6 +483,73 @@ def load(self, name: str, index: sympy.Expr):
...
+                const_term += coeff
+                continue
+            coeff_map[base] = coeff_map.get(base, sympy.Integer(0)) + coeff
+
+        if not sympy.simplify(const_term) == 0:
+            return None
+
+        indices: list[sympy.Expr] = []
+        remaining_map = dict(coeff_map)
+        for size, stride in zip(sizes, strides):
```
Comment (Skylion007) [inline torch/_inductor/codegen/cutedsl/cutedsl_kernel.py:533]:
```suggestion
        for size, stride in zip(sizes, strides, strict=True):
